### PR TITLE
Added batchnorm layer with tests 

### PR DESCRIPTION
**Pull request by ducha-aiki**
based on fixed PRs from Russell91 (base test) and ChenglongChen (implementation)

### CODE REVIEW

**Comment by weiliu89** [comment]
I think the code misses computing the second part of derivative over mean during back propagation.

File: src/caffe/layers/bn_layer.cpp [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+    caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, N_, C_, 1, Dtype(1),
+        batch_sum_multiplier_.cpu_data(), batch_mean_.cpu_data(), Dtype(0),
+        spatial_mean_.mutable_cpu_data());
+    caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, N_ * C_, H_ * W_, 1, Dtype(1),
+        spatial_mean_.cpu_data(), spatial_sum_multiplier_.cpu_data(), Dtype(0),
+        bottom_diff);
+
+    caffe_mul(buffer_blob_.count(), x_norm_.cpu_data()/*top_data*/, bottom_diff, bottom_diff);
+
```
Comment (weiliu89) [inline src/caffe/layers/bn_layer.cpp]: // I think we should add these following code between line 244 and 246. Correct me if I am wrong

// EX across spatial
caffe_cpu_gemv<Dtype>(CblasNoTrans, N_ * C_, H_ * W_, Dtype(1), bottom_diff, spatial_multiplier_.cpu_data(), Dtype(0), spatial_mean_.mutable_cpu_data());
// EX across batch
caffe_cpu_gemv<Dtype>(CblasNoTrans, N_, C_, Dtype(1), spatial_mean_.cpu_data(), batch_sum_multiplier_.cpu_data(), Dtype(0), batch_mean_.mutable_cpu_data());

caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, N_, C_, 1, Dtype(1), batch_sum_multiplier_.cpu_data(), batch_mean_.cpu_data(), Dtype(0), spatial_mean_.mutable_cpu_data());
caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, N_ * C_, H_ * W_, 1, Dtype(-1), spatial_mean_.cpu_data(), spatial_multiplier_.cpu_data(), Dtype(1), bottom_diff);

**Comment by weiliu89** [comment]
During backward, it misses step: Average 2 as shown in the image I attached.
Correct me if I am wrong

bn_eq

**Comment by ChenglongChen** [comment]
@weiliu89,
The second part in Eq (3) can be dropped as x_i - \mu has zero mean? As a result Average 2 is not necessary.

**Comment by weiliu89** [comment]
@ChenglongChen 
I think you are correct. Do you have any idea how to compute the derivatives over input x? I have no idea. Do you have any reference on how to do it?

**Comment by ducha-aiki** [comment]
Replaced by #1965

**Pull request closed without merging**